### PR TITLE
S3objectstore

### PR DESCRIPTION
**Pull request by SimonRichardson**
Introduces the s3 objectstore. The s3 objectstore allows controllers to move into a HA setup. File objectstores only allows singular controller nodes (non-HA), so in production setups it is expected that s3 should be used. There are still a few missing pieces to close out the objectstore epic:

 1. Customising bucket name during bootstrap.
 2. Removal of orphaned stale blobs in the filestores.
 3. Removal of old state code. 
 4. Update integration tests to use the right backend.
 5. Allow the usage of secrets from the secrets backend (that'll be a different epic).

We can't do 2, without first doing 1 in the list. This is because you want to be able to use the same AWS account with multiple controllers. Name squatting on the same bucket means we don't know if we can remove a model, as the controller might not own all of the directories.

The s3 objectstore code is relatively similar to the file based objectstore. Although there are some key differences, the s3 objectstore needs to know about the juju hashing algorithm; SHA384, and the s3 hashing algorithm; SHA256. Juju resources use SHA384 throughout and it might be wise at some point to move over to the more commonly used SHA256. Unfortunately, the import and export code will need to be modified to handle SHA256 instead of SHA384. Hence why I'm delaying doing that now.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

We're going to use local s3 setup, but this works just as well with AWS S3.

Install docker or podman (I'll demonstrate using docker).

```sh
docker run -d \
   -p 9000:9000 \
   -p 9001:9001 \
   --user $(id -u):$(id -g) \
   --name minio1 \
   -e "MINIO_ROOT_USER=ROOTUSER" \
   -e "MINIO_ROOT_PASSWORD=CHANGEME123" \
   -v ${HOME}/.local/minio/data:/data \
   quay.io/minio/minio server /data --console-address ":9001"
```

Go to http://localhost:9001/access-keys and add an access key, keeping note of the key and secret.

Locate the IP for minio, we'll use that later.

```sh
$ juju bootstrap lxd test --build-agent --config="object-store-type=s3" --config="object-store-s3-endpoint=http://${IP_ADDRESS}:9000" --config="object-store-s3-static-key=${S3_KEY}" --config="object-store-s3-static-secret=${S3_SECRET}"
```

Once deployed you can inspect minio bucket.

```sh
$ juju enable-ha
$ juju deploy juju-qa-test --resource foo-file="./tests/suites/resources/foo-file.txt"
```

## Links

**Jira card:** JUJU-5232

### CODE REVIEW

**Comment by jameinel** [comment]
Ideally we wouldn't have to define the s3 bucket at bootstrap vs being able to set it as needed. Obviously moving data from an old value to a new value is a difficulty, but we also want to support bootstrapping to local and then moving to s3, so we already need that functionality.

Is it reasonable to just use the controller-uuid as part of the key for the s3 bucket default, so that it always clearly owns its own space? That also would help avoid ever having collisions where a user asks 2 controllers to use the same bucket, and they start deleting each other's content.

**Comment by SimonRichardson** [comment]
> Ideally we wouldn't have to define the s3 bucket at bootstrap vs being able to set it as needed

That's next on the list after I've sorted the bucket naming out.

File: internal/objectstore/s3objectstore.go [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+type s3ObjectStore struct {
+	baseObjectStore
+	client    objectstore.Client
+	namespace string
+	requests  chan request
+}
+
+// NewS3ObjectStore returns a new object store worker based on the s3 backing
+// storage.
+func NewS3ObjectStore(ctx context.Context, namespace string, client objectstore.Client, metadataService objectstore.ObjectStoreMetadata, claimer Claimer, logger Logger, clock clock.Clock) (TrackedObjectStore, error) {
```
Comment (manadart) [inline internal/objectstore/s3objectstore.go:36]: Worth creating a struct for these args (not including context of course)?

File: internal/objectstore/s3objectstore.go [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+
+		// Now that we've written the file, we can upload it to the object
+		// store.
+		err := s.PutObject(ctx, defaultBucketName, t.filePath(fileEncodedHash), file, s3EncodedHash)
+		if err == nil {
+			return nil
+		}
+
+		// If the file already exists, then we can ignore the error.
+		if errors.Is(err, errors.AlreadyExists) {
```
Comment (manadart) [inline internal/objectstore/s3objectstore.go:343]: Could "or" this with the block above.

File: internal/s3client/client.go [hunk 3]
```diff
@@ -202,16 +222,33 @@ var forbiddenErrorCodes = map[string]struct{}{
 	"SignatureDoesNotMatch": {},
 }
 
+var alreadyExistCodes = map[string]struct{}{
```
Comment (manadart) [inline internal/s3client/client.go:225]: Hot enough path to have these on the heap forever? I know they're tiny...

File: internal/worker/units3caller/manifold.go [hunk 4]
```diff
@@ -107,3 +106,14 @@ func (c *httpClient) Do(req *http.Request) (*http.Response, error) {
 	err := c.client.Do(req.Context(), req, &res)
 	return res, err
 }
+
+// ensureHTTPS takes a URI and ensures that it is a HTTPS URL.
+func ensureHTTPS(address string) string {
```
Comment (manadart) [inline internal/worker/units3caller/manifold.go:111]: Seem to recall we do this for LXD. Worth making a utility function in core somewhere?

**Review by manadart — COMMENTED** [review]
Looks smashing based on semantic review. I will QA properly first thing.

File: internal/objectstore/base.go [hunk 5]
```diff
@@ -124,3 +177,16 @@ func (w *baseObjectStore) withLock(ctx context.Context, hash string, f func(cont
...
+
+type hashValidator func(string) (string, bool)
+
+func ignoreHash(string) (string, bool) {
+	return "", true
+}
+
+func checkHash(expected string) func(string) (string, bool) {
+	return func(actual string) (string, bool) {
+		loggo.GetLogger("***").Criticalf("checkHash: expected %q, actual %q", expected, actual)
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply internal/objectstore/base.go:189]: I did not, I'll remove. It was to help diagnose the two commit hash problem

File: internal/objectstore/s3objectstore.go (hunk 1 above)
Comment (SimonRichardson) [reply internal/objectstore/s3objectstore.go:36]: It's getting that way!

File: internal/worker/units3caller/manifold.go (hunk 4 above)
Comment (SimonRichardson) [reply internal/worker/units3caller/manifold.go:111]: I'll do that in the next PR

**Review by manadart — APPROVED** [review]
Looks great.

**Pull request merged**